### PR TITLE
Add encode callback url query option

### LINKED ISSUE #401: Double encoded callback url causes issues
### Description:

I've been using the framework successfully for awhile, but recently it stoped working with Imgur with `The redirect URI provided is missing or does not match` error. The issue doesn't seem to be related to any particular changes in the framework, but rather to Imgur backend updates. However, oauth works with Postman and Paw, which leads to certain conclusions.

Digging deeper the issue seems to be related to double encoded callback url in `OAuth2Swift::authorize` method. Removing double encoding solves the issue. I believe services take this into account and un-encode the url, which is good and bad, as it makes it easier to work with, but introduces uncertainty.

[`queryString` composed in `OAuth2Swift::authorize`](https://github.com/OAuthSwift/OAuthSwift/blob/b50109fd752f93c93151de4b7695bd710aa26bed/Sources/OAuth2Swift.swift#L121-L143) combines every parameter as is with exception for callback url, which gets encoded. The composed query then gets encoded, which causes url to be double encoded. If query would be composed with every parameter encoded immediately and without re-encoding everything, it would produce a valid query and eliminate the issue.

P. S. I see that logic [has changed recently](https://github.com/OAuthSwift/OAuthSwift/blob/b50109fd752f93c93151de4b7695bd710aa26bed/Sources/OAuth2Swift.swift#L122-L126), however, the problem occurs in the following `urlQueryEncoded` invocation.

### OAuth Provider (Twitter, Github, ..):
- [x] Imgur

### OAuth Version:
- [x] Version 2

### OS (Please fill the version) :
- [x] OSX :

### Installation method:
- [x] Carthage

### Library version:
- [x] v1.1.2

### Xcode version:
- [x] 8.0 (Swift 3.0)

### PR DESCRIPTION
**Pull request by ianbytchek**
This gives control over double encoding and fixes #401.

### CODE REVIEW

File: Sources/OAuth2Swift.swift [hunk 1]
```diff
@@ -118,32 +124,24 @@ open class OAuth2Swift: OAuthSwift {
...
-            queryString += "&\(param.0)=\(param.1)"
-        }
+        var queryString: String? = ""
 
-        var urlString = self.authorizeUrl
-        urlString += (self.authorizeUrl.contains("?") ? "&" : "?")
+        queryString = queryString?.urlQueryByAppending(parameter: "client_id", value: self.consumerKey)
+        queryString = queryString?.urlQueryByAppending(parameter: "redirect_uri", value: self.encodeCallbackURL ? callbackURL.absoluteString.urlEncoded : callbackURL.absoluteString, encode: self.encodeCallbackURLQuery)
+        queryString = queryString?.urlQueryByAppending(parameter: "response_type", value: self.responseType)
+        queryString = queryString?.urlQueryByAppending(parameter: "scope", value: scope)
```
Comment (phimage) [inline Sources/OAuth2Swift.swift]: some providers do not support empty parameters like scope, parameter, that`s why original code have `if !scope.isEmpty {` ` if !state.isEmpty {`

File: Sources/OAuth2Swift.swift [hunk 2]
```diff
@@ -290,3 +289,13 @@ open class OAuth2Swift: OAuthSwift {
     }
 
 }
+
+extension OAuth2Swift {
+
+    public enum CallbackURLEncode {
```
Comment (phimage) [inline Sources/OAuth2Swift.swift]: you choose the two boolean option but keep an unused enum

File: Sources/OAuth2Swift.swift [hunk 3]
```diff
@@ -118,32 +124,24 @@ open class OAuth2Swift: OAuthSwift {
...
+            queryString = queryString?.urlQueryByAppending(parameter: name, value: "\(value)")
+        }
+
+        if let queryString: String = queryString, let url: URL = URL(string: self.authorizeUrl.urlByAppending(query: queryString)) {
+            self.authorizeURLHandler.handle(url)
             return self
         } else {
             self.cancel() // ie. remove the observer.
-            failure?(OAuthSwiftError.encodingError(urlString: urlString))
+            failure?(OAuthSwiftError.encodingError(urlString: self.authorizeUrl))
```
Comment (phimage) [inline Sources/OAuth2Swift.swift]: here we loose information about what could not be encoded/converted to URL
maybe replace by `self.authorizeUrl.urlByAppending(query: queryString`

File: Sources/OAuth2Swift.swift [hunk 4]
```diff
@@ -118,32 +124,24 @@ open class OAuth2Swift: OAuthSwift {
...
-        if !scope.isEmpty {
-            queryString += "&scope=\(scope)"
-        }
-        if !state.isEmpty {
-            queryString += "&state=\(state)"
-        }
-        for param in parameters {
-            queryString += "&\(param.0)=\(param.1)"
-        }
+        var queryString: String? = ""
```
Comment (phimage) [inline Sources/OAuth2Swift.swift]: optional?

Comment (ianbytchek) [inline Sources/OAuth2Swift.swift]: To make this work, parameters are encoded as they are added, which may produce `nil` value, just like when you [encode the whole string](https://github.com/OAuthSwift/OAuthSwift/blob/b50109fd752f93c93151de4b7695bd710aa26bed/Sources/OAuth2Swift.swift#L141). Here we add them and if everything is ok, we end up with the same encoded query string, if not – it's `nil`, which we [check in the same fashion](https://github.com/ianbytchek/OAuthSwift/blob/3915fcace18ede9bf21d7767a16ac80df967db96/Sources/OAuth2Swift.swift#L139) and [fail with the same error](https://github.com/ianbytchek/OAuthSwift/blob/3915fcace18ede9bf21d7767a16ac80df967db96/Sources/OAuth2Swift.swift#L144).

File: Sources/OAuth2Swift.swift (hunk 1 above)
Comment (ianbytchek) [inline Sources/OAuth2Swift.swift]: That's fine, they are [checked in `urlQueryByAppending` and not added if empty](https://github.com/ianbytchek/OAuthSwift/blob/3915fcace18ede9bf21d7767a16ac80df967db96/Sources/String%2BOAuthSwift.swift#L30).

File: Sources/OAuth2Swift.swift (hunk 3 above)
Comment (ianbytchek) [inline Sources/OAuth2Swift.swift]: I don't think we do – [`urlString` is defined a couple lines before](https://github.com/OAuthSwift/OAuthSwift/blob/b50109fd752f93c93151de4b7695bd710aa26bed/Sources/OAuth2Swift.swift#L138) and only gets appended with `?` or `&` in order to make it ready for [concatenation with with query string below](https://github.com/OAuthSwift/OAuthSwift/blob/b50109fd752f93c93151de4b7695bd710aa26bed/Sources/OAuth2Swift.swift#L141). When failure gets invoked, the variable would only differ from the property by one extra character.

File: Sources/OAuth2Swift.swift (hunk 2 above)
Comment (ianbytchek) [inline Sources/OAuth2Swift.swift]: Ooops, forgot about it. I've tried using enum as suggested, but realised that naming might be an issue and would cause issues for users, who already use the initial property, so went back to boolean forgetting to remove the declaration. Well noticed.

Comment (ianbytchek) [inline Sources/OAuth2Swift.swift]: Fixed.

File: Sources/OAuth2Swift.swift (hunk 3 above)
Comment (phimage) [inline Sources/OAuth2Swift.swift]: yes, the previous code is also wrong (because of encodedQuery, same here if no queryString )

File: Sources/OAuth2Swift.swift (hunk 4 above)
Comment (phimage) [inline Sources/OAuth2Swift.swift]: ok

File: Sources/OAuth2Swift.swift (hunk 1 above)
Comment (phimage) [inline Sources/OAuth2Swift.swift]: ok
so there is a behaviour change for response_type , client_id, redirect_uri
but I don't think that this could add issue

**Comment by phimage** [comment]
thanks for the PR
I will do a test and merge when I have some times

File: Sources/OAuth2Swift.swift (hunk 3 above)
Comment (ianbytchek) [inline Sources/OAuth2Swift.swift]: I thought that it might be nice to use [`URLComponents`](https://developer.apple.com/documentation/foundation/urlcomponents) approach with custom encoding, but thought that it might be a radical addition for this PR. Using string extensions and doing low-level url assembly in authorisation method feels wrong. I'm happy to make another PR if you are happy with the approach.
